{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sample QC\n",
    "\n",
    "This notebook is a quick guide to how sketching algorithms can be used for sample QC. The particular emphasis is on getting from raw sequencing data to filtered reads, which we will then use in the next steps of our genomics workflow.\n",
    "\n",
    "> Note: many sketching algorithms can work with raw data, which contributes to the speed benefits of sketching. That being said, if we want more than just a quick and dirty analysis, it's always best to QC your data...\n",
    "\n",
    "We'll use the following sketching algorithms, implemented via some excellent open-source bioinformatics software:\n",
    "\n",
    "* MinHash\n",
    "* CountMin sketch\n",
    "* Bloom filter\n",
    "\n",
    "***\n",
    "\n",
    "## The data\n",
    "\n",
    "As mentioned in 4.1, we are looking at a suspected outbreak of *E.cloacae* in a hospital. We are using the sequence data collected during the [Reuter et al.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4001082/) study. The following table contains the accessions for the isolates we need, I just grabbed this from the [supplementary file](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4001082/bin/NIHMS58061-supplement-Supplementary_Online_Content.pdf) and then added the ENA experiment IDs.\n",
    "\n",
    "|isolate name|patient ID|ENA sample|ENA experiment|\n",
    "|-|-|-|-|\n",
    "EC1a|EC1|ERS184249|ERX168346|\n",
    "EC2a|EC2|ERS184250|ERX168347|\n",
    "EC2b|EC2|ERS184251|ERX168341|\n",
    "EC3a|EC3|ERS184252|ERX168340|\n",
    "EC4a|EC4|ERS184245|ERX168345|\n",
    "EC5a|EC5|ERS184246|ERX168339|\n",
    "EC6a|EC6|ERS184247|ERX168343|\n",
    "EC7a|EC7|ERS184248|ERX168344|\n",
    "\n",
    "These samples correspond to whole genome sequence data for *E.cloacae* isolates, collected from several patients in the hospital during the suspected outbreak.\n",
    "\n",
    "To save us the pain of waiting for fastq-dump to collect the data, I have already downloaded it. For future reference though, use `fastq-dump` (part of [sra-tools](https://github.com/ncbi/sra-tools)) to download this data from the ENA. We could have sketched this download data stream, but as we know that we need this data for our analysis it is best to have it stored on disk. You'll see later how we can sketch a data stream from a download, evaluate these sketches to see if the data is helpful for our analysis, and then decide to store or discard it.\n",
    "\n",
    "The sequence data for these isolates is actually pretty poor quality - [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) showed the terminal 50 bases to be full of Ns and the quality dropped below Q20. So I ran Trimmomatic to clean the data, here is the command ran for one sample:\n",
    "\n",
    "```\n",
    "trimmomatic PE ERX168346_1.fastq ERX168346_2.fastq ERX168346_1-trimmed.fq ERX168346_1-se ERX168346_2-trimmed.fq ERX168346_2-se SLIDINGWINDOW:5:20 MINLEN:100\n",
    "```\n",
    "\n",
    "So, we now have quality checked, trimmed sequence data for each sample downloaded and ready to go. This is stored in `../data/reads`. To make things easier to follow, the following steps will be ran on just one of these samples. Feel free to update the code to run on all of them!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sanity check\n",
    "\n",
    "Are my samples what I think they are?!\n",
    "\n",
    "The study labelled these isolates as *E.cloacae*, let's use **MinHash** to check. We'll be using the excellent [sourmash](https://github.com/dib-lab/sourmash) software. The sourmash docs actually have loads of great examples and workflows - be sure to check them out too.\n",
    "\n",
    "\n",
    "* start with downloading a reference database of MinHash genome sketches:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2019-04-08 13:42:49--  https://osf.io/4f8n3/download\n",
      "Resolving osf.io... 35.190.84.173\n",
      "Connecting to osf.io|35.190.84.173|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 FOUND\n",
      "Location: https://files.osf.io/v1/resources/vk4fa/providers/osfstorage/5c17c7754eb58100178fa5b3?action=download&version=1&direct [following]\n",
      "--2019-04-08 13:42:50--  https://files.osf.io/v1/resources/vk4fa/providers/osfstorage/5c17c7754eb58100178fa5b3?action=download&version=1&direct\n",
      "Resolving files.osf.io... 35.186.214.196\n",
      "Connecting to files.osf.io|35.186.214.196|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 125814027 (120M) [application/octet-stream]\n",
      "Saving to: 'genbank-k31.lca.json.gz'\n",
      "\n",
      "genbank-k31.lca.jso 100%[===================>] 119.99M  4.37MB/s    in 28s     \n",
      "\n",
      "2019-04-08 13:43:20 (4.28 MB/s) - 'genbank-k31.lca.json.gz' saved [125814027/125814027]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# download the pre-made MinHash database (courtesy of sourmash) containing all GenBank microbial genomes\n",
    "!wget -O genbank-k31.lca.json.gz https://osf.io/4f8n3/download"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* next, create a MinHash sketch of the reads from one sample:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[K== This is sourmash version 2.0.0. ==\n",
      "\u001b[K== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==\n",
      "\n",
      "\u001b[Ksetting num_hashes to 0 because --scaled is set\n",
      "\u001b[Kcomputing signatures for files: ../data/reads/ERX168346_1-trimmed.fq.gz, ../data/reads/ERX168346_2-trimmed.fq.gz\n",
      "\u001b[KComputing signature for ksizes: [31]\n",
      "\u001b[KComputing only nucleotide (and not protein) signatures.\n",
      "\u001b[KComputing a total of 1 signature(s).\n",
      "\u001b[K... reading sequences from ../data/reads/ERX168346_1-trimmed.fq.gz\n",
      "\u001b[K... ../data/reads/ERX168346_1-trimmed.fq.gz 125643 sequences\n",
      "\u001b[K... reading sequences from ../data/reads/ERX168346_2-trimmed.fq.gz\n",
      "\u001b[K... ../data/reads/ERX168346_2-trimmed.fq.gz 125643 sequences\n",
      "\u001b[Kcalculated 1 signatures for 251286 sequences taken from 2 files\n",
      "\u001b[Ksaved 1 signature(s). Note: signature license is CC0.\n"
     ]
    }
   ],
   "source": [
    "# sketch the reads from a sample\n",
    "!sourmash compute --scaled 1000 -k 31 ../data/reads/ERX168346_*-trimmed.fq.gz --merge ERX168346 -o ERX168346.sketch "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> the --scaled flag tells sourmash to decide the number of hashes to include in the MinHash sketch. Sourmash decides this based on the sequence length; it is effectively setting a compression ratio of 1000-to-1.\n",
    "\n",
    "> make sure that k-mer size for the sketch matches the k-mer size of the reference database (31)\n",
    "\n",
    "> the --merge flag tells sourmash to combine all the sketches into a single sketch. This is needed as our data is paired-end and we just want to have a single sketch per sample.\n",
    "\n",
    "* now we have sketched our sample, we can compare it to the reference sketches and check we have the organism we expect (e.cloacae!):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[K== This is sourmash version 2.0.0. ==\n",
      "\u001b[K== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==\n",
      "\n",
      "\u001b[Kloaded query: ERX168346... (k=31, DNA)\n",
      "\u001b[Kloaded 1 databases.                                                            \n",
      "\n",
      "\n",
      "overlap     p_query p_match\n",
      "---------   ------- -------\n",
      "4.4 Mbp       95.1%   91.7%    FKBP01000001.1 Enterobacter cloacae s...\n",
      "\u001b[Kfound less than 10.0 kbp in common. => exiting\n",
      "\n",
      "found 1 matches total;\n",
      "the recovered matches hit 95.1% of the query\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# compare the sketch of the reads against each sketch in the reference database\n",
    "!sourmash gather -k 31 ERX168346.sketch genbank-k31.lca.json.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So the sketch from our sample matches the sketch of the E.cloacae reference genome - yay!\n",
    "\n",
    "It wasn't a complete match however, there is a little bit of a difference between the sketches. This is probably due to sequencing error. We will have a fair few unique k-mers in the sequencing data, which typically arise from errors during sequencing (but could also be from strain variation).\n",
    "\n",
    "Let's use some more sketching to investigate!\n",
    "\n",
    "\n",
    "## K-mer spectrum\n",
    "\n",
    "First off, let's check that our guess is right. Do we have a lot of unique k-mers? To answer this, we'll use the **Count-Min sketch** and another piece of software from Titus Brown et al. - [khmer](https://github.com/dib-lab/khmer).\n",
    "\n",
    "* begin by using a Count-Min sketch to create a k-mer count graph (the k-mer spectrum):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "|| This is the script abundance-dist-single.py in khmer.\n",
      "|| You are running khmer version 3.0.0a2\n",
      "|| You are also using screed version 1.0\n",
      "||\n",
      "|| If you use this script in a publication, please cite EACH of the following:\n",
      "||\n",
      "||   * MR Crusoe et al., 2015. http://dx.doi.org/10.12688/f1000research.6924.1\n",
      "||   * Q Zhang et al., http://dx.doi.org/10.1371/journal.pone.0101271\n",
      "||   * A. Döring et al. http://dx.doi.org:80/10.1186/1471-2105-9-11\n",
      "||\n",
      "|| Please see http://khmer.readthedocs.io/en/latest/citations.html for details.\n",
      "\n",
      "\n",
      "PARAMETERS:\n",
      " - kmer size =     21 \t\t(-k)\n",
      " - n tables =      4 \t\t(-N)\n",
      " - max tablesize = 2.5e+08 \t(-x)\n",
      "Estimated memory usage is 1.0 Gb (1e+09 bytes = 4 bytes x 2.5e+08 entries / 1 entries per byte)\n",
      "--------\n",
      "making countgraph\n",
      "building k-mer tracking graph\n",
      "kmer_size: 21\n",
      "k-mer countgraph sizes: [215053697, 215053673, 215053669, 215053667]\n",
      "outputting to ERX168346.dist\n",
      "consuming input, round 1 -- ../data/reads/ERX168346_1-trimmed.fq.gz\n",
      "Total number of unique k-mers: 4289811\n",
      "preparing hist from ../data/reads/ERX168346_1-trimmed.fq.gz...\n",
      "consuming input, round 2 -- ../data/reads/ERX168346_1-trimmed.fq.gz\n",
      "wrote to: ERX168346.dist\n"
     ]
    }
   ],
   "source": [
    "!abundance-dist-single.py -M 1e9 -k 21 -s ../data/reads/ERX168346_1-trimmed.fq.gz ERX168346.dist"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* plot the k-mer abundance against k-mer count:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'inline%'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-5ee8554a3a15>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_line_magic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'matplotlib'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'inline%'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mnumpy\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mpylab\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mdist1\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnumpy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloadtxt\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'ERX168346.dist'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mskiprows\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdelimiter\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m','\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdist1\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdist1\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/genome-sketching/lib/python3.6/site-packages/IPython/core/interactiveshell.py\u001b[0m in \u001b[0;36mrun_line_magic\u001b[0;34m(self, magic_name, line, _stack_depth)\u001b[0m\n\u001b[1;32m   2305\u001b[0m                 \u001b[0mkwargs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'local_ns'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getframe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstack_depth\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf_locals\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2306\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuiltin_trap\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2307\u001b[0;31m                 \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2308\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2309\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m</anaconda3/envs/genome-sketching/lib/python3.6/site-packages/decorator.py:decorator-gen-108>\u001b[0m in \u001b[0;36mmatplotlib\u001b[0;34m(self, line)\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/genome-sketching/lib/python3.6/site-packages/IPython/core/magic.py\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(f, *a, **k)\u001b[0m\n\u001b[1;32m    185\u001b[0m     \u001b[0;31m# but it's overkill for just that one bit of state.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    186\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mmagic_deco\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 187\u001b[0;31m         \u001b[0mcall\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    188\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    189\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcallable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/genome-sketching/lib/python3.6/site-packages/IPython/core/magics/pylab.py\u001b[0m in \u001b[0;36mmatplotlib\u001b[0;34m(self, line)\u001b[0m\n\u001b[1;32m     97\u001b[0m             \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Available matplotlib backends: %s\"\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mbackends_list\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     98\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 99\u001b[0;31m             \u001b[0mgui\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbackend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshell\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0menable_matplotlib\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgui\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    100\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_show_matplotlib_backend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgui\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbackend\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    101\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/genome-sketching/lib/python3.6/site-packages/IPython/core/interactiveshell.py\u001b[0m in \u001b[0;36menable_matplotlib\u001b[0;34m(self, gui)\u001b[0m\n\u001b[1;32m   3370\u001b[0m         \"\"\"\n\u001b[1;32m   3371\u001b[0m         \u001b[0;32mfrom\u001b[0m \u001b[0mIPython\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcore\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mpylabtools\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mpt\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3372\u001b[0;31m         \u001b[0mgui\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbackend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfind_gui_and_backend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgui\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpylab_gui_select\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3373\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3374\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mgui\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m'inline'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/genome-sketching/lib/python3.6/site-packages/IPython/core/pylabtools.py\u001b[0m in \u001b[0;36mfind_gui_and_backend\u001b[0;34m(gui, gui_select)\u001b[0m\n\u001b[1;32m    278\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mgui\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mgui\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m'auto'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    279\u001b[0m         \u001b[0;31m# select backend based on requested gui\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 280\u001b[0;31m         \u001b[0mbackend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbackends\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mgui\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    281\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mgui\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'agg'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    282\u001b[0m             \u001b[0mgui\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'inline%'"
     ]
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "import numpy\n",
    "from pylab import *\n",
    "dist1 = numpy.loadtxt('ERX168346.dist', skiprows=1, delimiter=',')\n",
    "plot(dist1[:,0], dist1[:,1])\n",
    "axis(xmax=50)\n",
    "title('k-mer abundance histogram')\n",
    "xlabel('k-mer abundance')\n",
    "ylabel('N k-mers with that abundance')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see a a lot unique k-mers in that plot. One reason to remove these is to reduce the amount of data we process in downstream analysis, which will reduce analysis time and take less memory (as there are fewer unique k-mers to hold in memory). That being said, it's not really necessary and removing them can have a negative impact on some analyses as we would be reducing coverage.\n",
    "\n",
    "So instead of removing them, let's first try to correct the sequence reads."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read error correction using Bloom filters\n",
    "\n",
    "For this, we can use [lighter](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0509-9). We need a bit of information first:\n",
    "\n",
    "* what size is the genome we have sequenced\n",
    "* how many reads do we have\n",
    "\n",
    "This information is used to calculate the sequencing coverage, this is then used to set the sampling fraction (alpha) so that it is inversely proportion to the depth of sequencing.\n",
    "\n",
    "We saw in the output of our earlier khmer command that we have 125643 read pairs (read length ~140bp) for ERX168346. We also know that the genome size of *E.cloacae* is 5.31Mb. So our genome coverage from this sample is approximately 7X (which isn't great, but we did aggressively trim the data as the quality was poor).\n",
    "\n",
    "We calculate our sampling fraction for each file using the equation from the lighter paper: 7/coverage.\n",
    "\n",
    "This means our sampling fraction is 2.13\n",
    "\n",
    "* now let's run lighter:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2019-04-09 13:59:06] =============Start====================\n",
      "[2019-04-09 13:59:06] Bad quality threshold is \"5\"\n",
      "[2019-04-09 13:59:08] Finish sampling kmers\n",
      "[2019-04-09 13:59:08] Bloom filter A's false positive rate: 0.000730\n",
      "[2019-04-09 13:59:13] Finish storing trusted kmers\n",
      "[2019-04-09 13:59:17] Finish error correction\n",
      "Processed 125643 reads:\n",
      "\t115592 are error-free\n",
      "\tCorrected 4969 bases(0.494379 corrections for reads with errors)\n",
      "\tTrimmed 0 reads with average trimmed bases 0.000000\n",
      "\tDiscard 0 reads\n",
      "[2019-04-09 13:59:17] =============Start====================\n",
      "[2019-04-09 13:59:17] Bad quality threshold is \"5\"\n",
      "[2019-04-09 13:59:20] Finish sampling kmers\n",
      "[2019-04-09 13:59:20] Bloom filter A's false positive rate: 0.000728\n",
      "[2019-04-09 13:59:25] Finish storing trusted kmers\n",
      "[2019-04-09 13:59:28] Finish error correction\n",
      "Processed 125643 reads:\n",
      "\t115301 are error-free\n",
      "\tCorrected 5081 bases(0.491298 corrections for reads with errors)\n",
      "\tTrimmed 0 reads with average trimmed bases 0.000000\n",
      "\tDiscard 0 reads\n"
     ]
    }
   ],
   "source": [
    "!lighter -r ../data/reads/ERX168346_1-trimmed.fq.gz -k 17 5310000 2.13\n",
    "!lighter -r ../data/reads/ERX168346_2-trimmed.fq.gz -k 17 5310000 2.13"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> lighter needs a k-mer size, the genome size and the alpha. Alpha is optional, but if you don't include it lighter will do an extra pass of your data to calculate it for you.\n",
    "\n",
    "Lighter corrected around 5000 bases - great! Let's now do some low-abundance k-mer trimming.\n",
    "\n",
    "\n",
    "## K-mer trimming using Count-Min sketch\n",
    "\n",
    "Back to the Count-Min sketch and khmer...\n",
    "\n",
    "* perform k-mer frequency based trimming on the corrected reads from lighter:\n",
    "\n",
    "> this takes a little while"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "|| This is the script trim-low-abund.py in khmer.\n",
      "|| You are running khmer version 3.0.0a2\n",
      "|| You are also using screed version 1.0\n",
      "||\n",
      "|| If you use this script in a publication, please cite EACH of the following:\n",
      "||\n",
      "||   * MR Crusoe et al., 2015. http://dx.doi.org/10.12688/f1000research.6924.1\n",
      "||   * Q Zhang, S Awad, CT Brown, https://dx.doi.org/10.7287/peerj.preprints.890v1\n",
      "||\n",
      "|| Please see http://khmer.readthedocs.io/en/latest/citations.html for details.\n",
      "\n",
      "\n",
      "PARAMETERS:\n",
      " - kmer size =     32 \t\t(-k)\n",
      " - n tables =      4 \t\t(-N)\n",
      " - max tablesize = 2e+09 \t(-x)\n",
      "Estimated memory usage is 8.0 Gb (8e+09 bytes = 4 bytes x 2e+09 entries / 1 entries per byte)\n",
      "--------\n",
      "making countgraph\n",
      "created temporary directory ./tmpduxdykibkhmer;\n",
      "use -T to change location\n",
      "... ERX168346_1-trimmed.cor.fq.gz 10015 10107 1239533 83 9976\n",
      "... ERX168346_1-trimmed.cor.fq.gz 19413 20001 2446816 571 68473\n",
      "... ERX168346_1-trimmed.cor.fq.gz 28563 30006 3665336 1419 170335\n",
      "... ERX168346_1-trimmed.cor.fq.gz 37536 40015 4886299 2447 293567\n",
      "... ERX168346_1-trimmed.cor.fq.gz 46358 50003 6098976 3608 432762\n",
      "... ERX168346_1-trimmed.cor.fq.gz 55230 60006 7318232 4736 568116\n",
      "... ERX168346_1-trimmed.cor.fq.gz 64084 70003 8534131 5873 704911\n",
      "... ERX168346_1-trimmed.cor.fq.gz 72934 80027 9758355 7042 846548\n",
      "... ERX168346_1-trimmed.cor.fq.gz 81710 90003 10971866 8236 988838\n",
      "... ERX168346_1-trimmed.cor.fq.gz 90509 100002 12187080 9430 1132315\n",
      "... ERX168346_1-trimmed.cor.fq.gz 99346 110010 13406243 10597 1272748\n",
      "... ERX168346_1-trimmed.cor.fq.gz 108060 120004 14624685 11871 1425619\n",
      "ERX168346_1-trimmed.cor.fq.gz: kept aside 112958 of 125643 from first pass\n",
      "... ERX168346_2-trimmed.cor.fq.gz 121888 135661 16521477 13690 1643143\n",
      "... ERX168346_2-trimmed.cor.fq.gz 130537 145644 17724481 15018 1801605\n",
      "... ERX168346_2-trimmed.cor.fq.gz 139231 155649 18931259 16324 1957106\n",
      "... ERX168346_2-trimmed.cor.fq.gz 147823 165650 20138681 17721 2124167\n",
      "... ERX168346_2-trimmed.cor.fq.gz 156332 175646 21345438 19200 2300975\n",
      "... ERX168346_2-trimmed.cor.fq.gz 164901 185649 22551971 20617 2470430\n",
      "... ERX168346_2-trimmed.cor.fq.gz 173418 195646 23756732 22084 2645720\n",
      "... ERX168346_2-trimmed.cor.fq.gz 181851 205670 24965193 23656 2834548\n",
      "... ERX168346_2-trimmed.cor.fq.gz 190217 215646 26167188 25252 3023897\n",
      "... ERX168346_2-trimmed.cor.fq.gz 198511 225645 27373999 26940 3225904\n",
      "... ERX168346_2-trimmed.cor.fq.gz 206760 235653 28581367 28681 3433879\n",
      "... ERX168346_2-trimmed.cor.fq.gz 214835 245646 29788597 30572 3659395\n",
      "ERX168346_2-trimmed.cor.fq.gz: kept aside 106377 of 125643 from first pass\n",
      "second pass: looking at sequences kept aside in ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2\n",
      "... x 2 125644 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 251287 30467935 31699 3793491\n",
      "... x 2 125645 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 251288 30468074 31700 3793620\n",
      "... x 2 125646 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 251289 30468177 31701 3793759\n",
      "... x 2 125647 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 251290 30468308 31702 3793862\n",
      "... x 2 125648 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 251291 30468426 31703 3793993\n",
      "... x 2 125649 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 251292 30468526 31704 3794111\n",
      "... x 2 125650 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 251293 30468666 31705 3794211\n",
      "... x 2 125651 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 251294 30468806 31706 3794351\n",
      "... x 2 125652 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 251295 30468945 31707 3794491\n",
      "... x 2 125653 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 251296 30469054 31708 3794630\n",
      "... x 2 125654 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 251297 30469177 31709 3794739\n",
      "... x 2 125655 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 251298 30469317 31710 3794862\n",
      "... x 2 130001 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 255644 31005462 36048 4328870\n",
      "... x 2 140001 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 265644 32225348 46025 5544925\n",
      "... x 2 150001 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 275644 33446261 55992 6760657\n",
      "... x 2 160001 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 285644 34664443 65972 7975076\n",
      "... x 2 170001 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 295644 35885313 75942 9190914\n",
      "... x 2 180001 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 305644 37103612 85926 10405601\n",
      "... x 2 190001 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 315644 38321755 95904 11619513\n",
      "... x 2 200001 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 325644 39543382 105883 12837421\n",
      "... x 2 210001 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 335644 40761879 115856 14051510\n",
      "... x 2 220001 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 345644 41979898 125832 15265434\n",
      "... x 2 230001 ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2 219335 355644 43200243 135803 16480900\n",
      "removing ./tmpduxdykibkhmer/ERX168346_1-trimmed.cor.fq.gz.pass2\n",
      "second pass: looking at sequences kept aside in ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2\n",
      "... x 2 238602 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364245 44249193 144377 17525252\n",
      "... x 2 238603 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364246 44249324 144378 17525366\n",
      "... x 2 238604 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364247 44249457 144379 17525497\n",
      "... x 2 238605 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364248 44249580 144380 17525630\n",
      "... x 2 238606 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364249 44249703 144381 17525753\n",
      "... x 2 238607 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364250 44249832 144382 17525876\n",
      "... x 2 238608 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364251 44249941 144383 17526005\n",
      "... x 2 238609 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364252 44250070 144384 17526114\n",
      "... x 2 238610 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364253 44250208 144385 17526243\n",
      "... x 2 238611 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364254 44250330 144386 17526381\n",
      "... x 2 238612 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364255 44250444 144387 17526503\n",
      "... x 2 238613 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364256 44250575 144388 17526617\n",
      "... x 2 238614 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364257 44250706 144389 17526748\n",
      "... x 2 238615 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364258 44250819 144390 17526879\n",
      "... x 2 238616 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364259 44250919 144391 17526992\n",
      "... x 2 238617 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364260 44251027 144392 17527092\n",
      "... x 2 238618 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364261 44251160 144393 17527200\n",
      "... x 2 238619 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364262 44251291 144394 17527333\n",
      "... x 2 238620 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364263 44251396 144395 17527464\n",
      "... x 2 238621 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364264 44251508 144396 17527569\n",
      "... x 2 238622 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364265 44251641 144397 17527681\n",
      "... x 2 238623 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364266 44251751 144398 17527814\n",
      "... x 2 238624 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 364267 44251861 144399 17527924\n",
      "... x 2 240001 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 365644 44417665 145772 17692862\n",
      "... x 2 250001 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 375644 45626636 155741 18896882\n",
      "... x 2 260001 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 385644 46833338 165719 20100004\n",
      "... x 2 270001 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 395644 48039882 175702 21303611\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... x 2 280001 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 405644 49249013 185690 22510558\n",
      "... x 2 290001 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 415644 50455849 195670 23713901\n",
      "... x 2 300001 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 425644 51661712 205652 24916749\n",
      "... x 2 310001 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 435644 52867550 215635 26119321\n",
      "... x 2 320001 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 445644 54075187 225621 27324426\n",
      "... x 2 330001 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 455644 55283125 235599 28528777\n",
      "... x 2 340001 ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2 219335 465644 56492056 245586 29735673\n",
      "removing ./tmpduxdykibkhmer/ERX168346_2-trimmed.cor.fq.gz.pass2\n",
      "removing temp directory & contents (./tmpduxdykibkhmer)\n",
      "read 251286 reads, 30467806 bp\n",
      "wrote 250557 reads, 30334920 bp\n",
      "looked at 219335 reads twice (1.87 passes)\n",
      "removed 729 reads and trimmed 1953 reads (1.07%)\n",
      "trimmed or removed 0.44%% of bases (132886 total)\n",
      "251286 reads were high coverage (100.00%);\n",
      "skipped 0 reads/0 bases because of low coverage\n",
      "fp rate estimated to be 0.000\n",
      "output in ERX168346.corrected.trimmed.fq\n"
     ]
    }
   ],
   "source": [
    "!trim-low-abund.py -V -M 8e9 -C 3 -Z 10 -o ERX168346.corrected.trimmed.fq ERX168346_1-trimmed.cor.fq.gz ERX168346_2-trimmed.cor.fq.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* now, plot the k-mer spectrum again now the reads have been corrected and k-mer trimmed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "|| This is the script abundance-dist-single.py in khmer.\n",
      "|| You are running khmer version 3.0.0a2\n",
      "|| You are also using screed version 1.0\n",
      "||\n",
      "|| If you use this script in a publication, please cite EACH of the following:\n",
      "||\n",
      "||   * MR Crusoe et al., 2015. http://dx.doi.org/10.12688/f1000research.6924.1\n",
      "||   * Q Zhang et al., http://dx.doi.org/10.1371/journal.pone.0101271\n",
      "||   * A. Döring et al. http://dx.doi.org:80/10.1186/1471-2105-9-11\n",
      "||\n",
      "|| Please see http://khmer.readthedocs.io/en/latest/citations.html for details.\n",
      "\n",
      "\n",
      "PARAMETERS:\n",
      " - kmer size =     21 \t\t(-k)\n",
      " - n tables =      4 \t\t(-N)\n",
      " - max tablesize = 2.5e+08 \t(-x)\n",
      "Estimated memory usage is 1.0 Gb (1e+09 bytes = 4 bytes x 2.5e+08 entries / 1 entries per byte)\n",
      "--------\n",
      "making countgraph\n",
      "building k-mer tracking graph\n",
      "kmer_size: 21\n",
      "k-mer countgraph sizes: [215053697, 215053673, 215053669, 215053667]\n",
      "outputting to ERX168346.dist\n",
      "consuming input, round 1 -- ERX168346.corrected.trimmed.fq\n",
      "Total number of unique k-mers: 4628648\n",
      "preparing hist from ERX168346.corrected.trimmed.fq...\n",
      "consuming input, round 2 -- ERX168346.corrected.trimmed.fq\n",
      "wrote to: ERX168346.dist\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, 'N k-mers with that abundance')"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZsAAAEWCAYAAACwtjr+AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAIABJREFUeJzt3XmcHHWd//HXe+5kMpNkchFykGAicihIQojijUI8YRU0rkrUKIvLKv7UXUFdQQ5X1lW8WRGQQ1zI4kEWQcgGEFfDERC5AiQkgcScZIZkck5Pz+f3R32bVIaZnppMV3dP+vN8PPrRXd+q+ta3a2rmM99vfev7lZnhnHPOpamq1AVwzjl34PNg45xzLnUebJxzzqXOg41zzrnUebBxzjmXOg82zjnnUufBxhWcpNWS3l7qcuwPSSZpWpGOVVbnSdI1ki7Os367pEOLWSZ34PBg45xLxMyGmdnKfNtIeouktcUqkxs8PNi4A5akmlKXwRWWpOpSl8HtHw82LlWSXiVplaS5vay/QNJ/S/qFpHZJj0l6paTzJG2StEbSSbHth0u6StJ6SX+TdHHuD5Ckj0v6k6TLJLUCF/RwvFmSlkh6MeTxI0l13TZ7l6SVkl6Q9G1JVbGy/iKW15TQ7FYTlu+RdFEoQ7ukOyWNjm3/MUnPSdoi6av9KVc4zlmSlktqk/RjSYqt/7SkZeG4T0o6NqQfLOlXkjaHn8Pn+viRjZT0u5DP/ZJe0a0M08Lnd4XjtIefw5ckNQK3AweHJrft4fj1kr4naV14fU9SfSzffwnfeZ2kT3U7zjWSLpd0m6QdwFslvVvSXyRtC9fHBbG8cj+TT4R1beG8HSfp0XB+f9THOXBpMDN/+augL2A18HbgWOB54D15tr0A2A2cDNQA1wGrgK8CtcCngVWx7X8L/BRoBMYCDwD/ENZ9HOgEPhvyGtLD8WYAs8P6KcAy4POx9QbcDbQAk4FngE/FyvqL2LZTwvY1Yfke4FnglcCQsPytsO4IYDvwJqAe+G4o69v7Ua5bgRGhXJuBOWHd6cDfgOMAAdOAQ4j+mXwI+DpQBxwKrARO7uVncQ3QCswK5bgBuLFbGaaFz+uBN4bPI4Fjw+e3AGu75XshcF/4eY0B/gxcFNbNATYARwJDgeu7HecaYCtwQvg+DeEYrw7LrwE2Aqd2+5n8Z9j2JKLr67fh+BOATcCbS/17UmmvkhfAXwfeiyjYfANYC7y1j20vABbFlt8b/ihXh+Wm8MdjBDAO2EMsiAAfBu4Onz8OPN/Psn4e+E1s2XJ/xMPyPwKLY2XtK9h8rdu+vw+fv97tD3cj0EEINgnL9YbY8gLg3PD5DuCcHvI4vvv5AM4Dft7LMa8Browtvwt4qlsZckHgeeAfgOZuebyFlwebZ4F3xZZPBlaHz1cD/xZbN42XB5vr+vgZfg+4rNvPZEJs/RbgQ7HlXxEL5P4qzsub0VxazgL+bGZ35xIkfSTWvHJ7bNuNsc+7gBfMLBtbBhhG9N96LbA+NIe8SFTLGRvbf02+QoUmulslbZC0DfgmMLrbZvE8ngMOzvtN97Uh9nlnKDchj5fyNbMdRH8E+1Ou3vKeRPQHvbtDiJq0Xoydr68QBe3+lr+7DxAFo+ck/UHS6/LkeTDRecyJn9N9zgs9//z2SZN0vKS7Q9PgVqJrrfu56n5NdV/u7Xu5lHiwcWk5C5gs6bJcgpndYFGPpmFm9s79yHMNUc1mtJmNCK9mMzsytk1fw5hfDjwFTDezZqI/vuq2zaTY58nAuvB5B1FTT85B/Sj7+ni+koYCo/pZrt6sAV7RS/qq2LkaYWZNZvaufpS7R2b2oJmdQhTof0tU04Kez/86osCXEz+n64GJsXXxc//S4bot/xJYCEwys+FETWZJz5UrEQ82Li3tRO3xb5L0rUJkaGbrgTuB70hqllQl6RWS3tyPbJqAbcB2Sa8CPtPDNv8saaSkScA5wE0h/RGi7zNZ0nCiJqmkbgbeI+kN4cb/hez7+5ekXL25EviSpBmKTJN0CNH9rG2SvixpiKRqSUdJOq4feb+MpLpQSx1uZplQ7lxNdCMwKpyfnP8CviZpjKIOE18Hch0tFgCfkHR4CMBfT1CEJqDVzHZLmgX8/UC+jysODzYuNWb2IvAO4J2SLipQtmcQ3ex+Emgj+iM+vh/7f4noj1M78DP2BpK4W4hurD8C/A64CsDMFoXtHw3rb016UDN7Ajib6L/y9aHs8edRkpSrt7z/G7gk5N1OVNNoCU2R7wWOIep08QJRYBreS1b98TFgdWjyOwv4aCjLU0TBZWVoujsYuBhYSnTeHgMeDmmY2e3AD4g6ZawAloT89+Q59j8CF0pqJwpOC/Js68qEwg0z55wrOUmHA48D9WbWWeryuMLxmo1zrqQk/V1omhsJXAr8jweaA48HG+dcqf0D0XNDzxLd++nP/So3SHgzmnPOudR5zcY551zqfKDCYPTo0TZlypRSF8M55waVhx566AUzG9PXdh5sgilTprB06dJSF8M55wYVSc/1vZU3oznnnCsCDzbOOedS58HGOedc6jzYOOecS50HG+ecc6nzYOOccy51Hmycc86lzoONc2Uq22Vcf99zPLi6la4uH1bKDW7+UKdzZerh59v4198+DsCYpnpOPnIc7371wRw/tYWqKp+Y0g0uHmycK1Nbtkfzh33xHa9k2YZt/Oqhv/GL+55nwoghfGDGRE47diKTRw3tIxfnyoMHG+fKVOuODACnz5zEQcMb2NWR5c4nN3DzQ2v54V3L+eFdy7n5rNcz45CRJS6pc33zezbOlam2nR0AjBhaC8CQumpOOWYC188/nts+90bMYNn6baUsonOJebBxrky17uigsa6ahtrql6175bgmqqvEhq27S1Ay5/rPg41zZaptRwcjG+t6XFddJcY11bNu664il8q5/ePBxrky1bqzg5Zegg3AQcMbvGbjBg0PNs6VqbYdHYwc2nuwGT98iAcbN2h4sHGuTPVVsxk/vIF1W3dh5g98uvLnwca5MtW2I5O3ZnPQ8AZ2Z7rYuitTxFI5t3882DhXhvZ0Ztm+p5OWxtpetzl4xBAA1r3oTWmu/Hmwca4Mvbgzqq301hsNopoNwIZt3iPNlT8PNs6VodYd0QOdLXk7CETBZr13EnCDQKrBRtIISTdLekrSMkmvk9QiaZGk5eF9ZGz78yStkPS0pJNj6TMkPRbW/UCSQnq9pJtC+v2SpsT2mReOsVzSvDS/p3OF1haCTb6azdimBqqrxHpvRnODQNo1m+8DvzezVwFHA8uAc4HFZjYdWByWkXQEMBc4EpgD/ERS7tHpy4EzgenhNSekzwfazGwacBlwacirBTgfOB6YBZwfD2rOlbvWMFRNvt5o1VVibFO912zcoJBasJHUDLwJuArAzDrM7EXgFODasNm1wKnh8ynAjWa2x8xWASuAWZLGA81mtsSiPp7Xddsnl9fNwImh1nMysMjMWs2sDVjE3gDlXNl7qWaTpxkNoqa09T6KgBsE0qzZHApsBn4u6S+SrpTUCIwzs/UA4X1s2H4CsCa2/9qQNiF87p6+zz5m1glsBUblyWsfks6UtFTS0s2bNw/kuzpXULkRn3ODcPbGH+x0g0WawaYGOBa43MxeC+wgNJn1oqfZoCxP+v7uszfB7Aozm2lmM8eMGZOnaM4VV9vODpobaqitzv8retDwBtZv3e0Pdrqyl2awWQusNbP7w/LNRMFnY2gaI7xvim0/Kbb/RGBdSJ/YQ/o++0iqAYYDrXnycm5QaN2Rf/SAnPHDG9iVyfqDna7spRZszGwDsEbSYSHpROBJYCGQ6x02D7glfF4IzA09zKYSdQR4IDS1tUuaHe7HnNFtn1xepwF3hfs6dwAnSRoZOgacFNKcGxTadvY+4nPc+OHRg53eScCVu7Rn6vwscIOkOmAl8AmiALdA0nzgeeB0ADN7QtICooDUCZxtZtmQz2eAa4AhwO3hBVHng+slrSCq0cwNebVKugh4MGx3oZm1pvlFnSuk1h0dHNTc0Od240fknrXZxeHjm9MulnP7LdVgY2aPADN7WHViL9tfAlzSQ/pS4Kge0ncTglUP664Gru5PeZ0rF207OhIFD3+w0w0WPoKAc2WorxGfc8YMq6dKeI80V/Y82DhXZnZ1ZNmd6erzGRuAmuoqxjU3+GCcrux5sHGuzOwdPSD/MzY5Bw1v8ME4XdlLHGzCA5nOuZQlHT0g5+DhQ3x8NFf2+gw2kl4v6Umicc2QdLSkn6ReMucq1EsjPie4ZwP+YKcbHJLUbC4jGmtsC4CZ/ZVozDPnXAradvY94nNc7sHObbs60yyWcwOSqBnNzNZ0S8r2uKFzbsCSzGUTl3uwc50PyOnKWJJgs0bS6wGTVCfpS4QmNedc4bXt6KBK0DwkeQcB8O7PrrwlCTZnAWezd/TlY8Kycy4FrTs7GDG0juqqnsaTfbmDwygCXrNx5azPEQTM7AXgI0Uoi3MOaNuRYWQfUwvE+YOdbjBI0hvtWkkjYssjJfkwMM6lJOmIzzm5Bzt9yBpXzpI0o70mzLAJQJj58rXpFcm5yta2syPxMzY5B/mMna7MJQk2VWGYfgAktZD+aNHOVaz+1mwADh4xhL+1ebBx5StJ0PgO8GdJN4fl0+lhZGbn3MCZWeK5bOKmjmrk949vIJPt6nN2T+dKoc+r0syuI5qYbCPRrJrvN7Pr0y6Yc5Vo+55OMllL/IxNzpTRjWS7jDWtO1MqmXMDk7Q57CmgLbe9pMlm9nxqpXKuQrXtiKZ37nfNZnQ0dOGqF3Zw6JhhBS+XcwPVZ7CR9FngfKKaTRYQYMBr0i2ac5WnvyM+5xwaCzbOlaMkNZtzgMPMbEvahXGu0vV3xOeckY11jBha68HGla1Ew9UAW9MuiHOu/yM+x00d3ejBxpWtJDWblcA9kn4H7Mklmtl3UyuVcxWqvyM+x00d1ciSld4A4cpTkprN88AioA5oir2ccwXWuqODmirRVN//R9mmjm5k/dbd7OrwQdld+UkyNto39jdzSauBdqKOBZ1mNjM8FHoTMAVYDXwwjEqApPOA+WH7z5nZHSF9BnANMAS4DTjHzExSPXAdMINovp0PmdnqsM884GuhKBeb2bX7+z2cK5bcMzZSskE446aOiToJrN6yg8PHNxe6aM4NSJKx0cZI+rak2yTdlXv14xhvNbNjzGxmWD4XWGxm04HFYRlJRwBzgSOBOcBPJFWHfS4HzgSmh9eckD4faDOzaUSTvF0a8moh6kF3PDALOD8+CoJz5ap1R0e/n7HJmeo90lwZS9KMdgPRczZTgW8Q1UYeHMAxTwFytYxrgVNj6Tea2R4zWwWsAGZJGg80m9kSi+a9va7bPrm8bgZOVPQv4cnAIjNrDbWmRewNUM6VrfbdnTQP2b/RoKaM8mDjyleSYDPKzK4CMmb2BzP7JDA7Yf4G3CnpIUlnhrRxZrYeILyPDekTiHq+5awNabl5dLqn77OPmXUS9ZoblScv58paJttFXc3+DTfTWF/DuOZ6Vm72YOPKT5J/oTLhfb2kdwPrgIkJ8z/BzNZJGgsskvRUnm17aqS2POn7u8/eA0YB8EyAyZMn5ymac8XRkTUa6/d/bLOpoxtZvcWDjSs/Sa7qiyUNB74IfAm4Evh/STI3s3XhfRPwG6L7JxtD0xjhfVPYfC0wKbb7RKLAtpZ9g1sufZ99JNUAw4HWPHl1L98VZjbTzGaOGTMmyVdyLlWZzi5qqgYSbIZ5M5orS0kG4rzVzLaa2eNm9lYzm2FmC/vaT1KjpKbcZ+Ak4HFgITAvbDYPuCV8XgjMlVQvaSpRR4AHQlNbu6TZ4X7MGd32yeV1GnBXuK9zB3BSmOhtZDj2HX2V2blSi5rR+t8TLefQ0Y207ujgxfC8jnPlotdmNEk/pIempxwz+1wfeY8DfhO6cNYAvzSz30t6EFggaT7RMzynh/yekLQAeBLoBM42s9wDA59hb9fn28ML4CrgekkriGo0c0NerZIuYm9HhgvNrLWP8jpXcgOdImBKrEfaayfvX68259KQ757N0vB+AnAE0bMxEAWHh/rK2MxWAkf3kL4FOLGXfS6hh7lyzGwpcFQP6btDeXrK62rAp692g0omawMKNlP3CTbe29+Vj16DTe4hSEkfJ3pWJhOW/xO4syilc67CdAywZjO5ZShVgtV+38aVmSRX9cHsOzzNsJDmnCuwTLaLuur9v2dTV1PFpJahrPRg48pMkq7P3wL+IunusPxm4ILUSuRcBescYDMa+OjPrjwlGRvt55JuJxr6BeBcM9uQbrGcq0wd2S5q9/Ohzpwpoxp5YFUrZrZfY6w5l4akV3U1sJloauhXSnpTekVyrjKZ2YB7owEcOqaRnR1ZNrXv6Xtj54okybTQlwIfAp4AukKyAfemWC7nKk62yzCD2qqB1UbiPdLGNTcUomjODViSezanEk0L7f8mOZeiTDZ6rG2gzWiHtETB5vnWncw+dNSAy+VcISS5qlcCtWkXxLlK15GNGg4G2ow2trkegE3bdg+4TM4VSpKazU7gEUmL2Xda6L5GEHDO9UMmBJuBdH0GaKitZsTQWjZu88YIVz6SBJuF4eWcS1GmQDUbgHFNDWz0mo0rI0m6Pvt0ys4VQaYz3LMpRLAZ7sHGlZckvdFW0cOAnGZ2aColcq5CvXTPZoAdBADGNdXzzIb2AefjXKEkaUabGfvcQDTwZUs6xXGucnV2FeaeDcC45gY2b99DtsuoHmBXaucKIcl8Nltir7+Z2feAtxWhbM5VlEI3o2W7jC3bvZOAKw9JmtGOjS1WEdV0mnrZ3Dm3nwrV9RmiZjSAjdv2MNYf7HRlIEkz2ndinzuB1cAHUymNcxUs1xutpkDNaAAbtu3m1QwfcH7ODVSS3mhvLUZBnKt0e5+zGXjN5qDhUbDxHmmuXPR5VUsaJekHkh6W9JCk70vyMTCcK7BCPmczqrGOKvkoAq58JLmqbyQa8fkDwGnh801593DO9VtHATsI1FRXMXpYPRs82LgykeSeTYuZXRRbvljSqWkVyLlK9VIzWk1huiofNLzBh6xxZSPJv1B3S5orqSq8Pgj8Lu2COVdpCtmMBjDWh6xxZaTXmo2kdqKRAwR8AfhFWFUFbAfOT710zlWQzmzhmtEAxjXX89BzrQXJy7mB6vWqNrMmM2sO71VmVhNeVWbWnPQAkqol/UXSrWG5RdIiScvD+8jYtudJWiHpaUknx9JnSHosrPuBwly3kuol3RTS75c0JbbPvHCM5ZLm9e+0OFd8hXzOBuCg5gbadmbY05ktSH7ODUSiq1rSSEmzJL0p9+rHMc4BlsWWzwUWm9l0YHFYRtIRwFzgSGAO8BNJ1WGfy4EzgenhNSekzwfazGwacBlwacirhajmdTwwCzg/HtScK0eF7PoMe5+12eT3bVwZSNL1+VNEU0DfAXwjvF+QJHNJE4F3A1fGkk8BciNJX0s0E2gu/UYz22Nmq4AVwCxJ44FmM1tiZgZc122fXF43AyeGWs/JwCIzazWzNmARewOUc2WpkA91QjRkDfizNq48JPkX6hzgOOC58IDna4m6PyfxPeBfgK5Y2jgzWw8Q3seG9AnAmth2a0PahPC5e/o++5hZJ7AVGJUnr31IOlPSUklLN29O+pWcS0cmhXs2gPdIc2UhyVW928x2Q3SPxMyeAg7raydJ7wE2mdlDCcvS079zlid9f/fZm2B2hZnNNLOZY8aMSVhM59LR0Zm7Z1Ogmk2T12xc+UjynM1aSSOA3wKLJLUB6xLsdwLwPknvIpqaoFnSL4CNksab2frQRLYpdxxgUmz/ieE4a8Pn7unxfdZKqgGGA60h/S3d9rknQZmdK5lMtovaahH6vwzYiKG11NVUebBxZSHJFAN/Z2YvmtkFwL8CV7H3nkm+/c4zs4lmNoXoxv9dZvZRoimmc73D5gG3hM8Lgbmhh9lUoo4AD4SmtnZJs8P9mDO67ZPL67RwDCO6r3RS6NgwEjgppDlXtqJgU5gmNABJjGuu92DjykKSms1LzOwPBTjmt4AFkuYDzxNNxoaZPSFpAfAk0ejSZ5tZrs/mZ4BrgCHA7eEFUeC7XtIKohrN3JBXq6SLgAfDdheamT9w4MpaJmsFDTYQNaX5PRtXDvoVbPaXmd1DaMYysy3Aib1sdwlwSQ/pS4GjekjfTQhWPay7Grh6f8vsXLF1FLhmA1GPtGXrthU0T+f2R2GvbOfcfuvMdhVkSui4cT5kjSsTSZ6zuTRJmnNuYDJZo7amwDWb5np2dGTZvqezoPk6119Jrux39JD2zkIXxLlK15HtoqaqsDWb3CRqG7Z67caVVq/BRtJnJD0GHCbp0dhrFfBo8YroXGXIdBb+ns3YptyQNR5sXGnl6yDwS6JeX/9GGL8saPeeXc4VXibbRV0KzWgAG9s92LjS6jXYmNlWouFfPgwgaSzRw5nDJA0zs+eLU0TnKkMqXZ+bc81o3v3ZlVaSDgLvlbQcWAX8AVjN3udcnHMF0hFGECikxvoamuprvEeaK7kk/0ZdDMwGnjGzqUTPyPwp1VI5V4EKPYJAztjmejZ5M5orsSRXdiY8iFklqcrM7gaOSblczlWcTLarYHPZxB00vMF7o7mSSzKCwIuShhHNaXODpE1Ew8k45woo01n4ezYQ3bdZ8uyWgufrXH8kubJPAXYB/w/4PfAs8N40C+VcJcp0dRX8oU6ASSOHsmHb7pemMHCuFPqs2ZjZjtjitb1u6JwbkEwKHQQAJrUMxQz+9uIupo5uLHj+ziWRpDfa+yUtl7RV0jZJ7ZJ8ZD/nCizTadRWpVGzGQLAmtadBc/buaSS3LP5d+C9ZrYs7cI4V8ky2S5qawpfs5k8aigAa9o82LjSSfJv1EYPNM6lL40pBiAa+bmuuoo1rbsKnrdzSfVas5H0/vBxqaSbiKaFfukxZDP7dcplc66ipNX1uapKTBg5xJvRXEnla0aL9zjbSTS1co4BHmycK6A0hqvJmThyiDejuZLKNzbaJwAknWBm+4wYIOmEtAvmXCXJdhnZrvSCzeSWodz22PpU8nYuiSRX9g8Tpjnn9lMmGz0Dk0YHAYi6P7ftzNC+O5NK/s71Jd89m9cBrwfGSPpCbFUzUJ12wZyrJLlgk8Y9G4ge7ARY07qLIw6uTeUYzuWT78quA4YRBaSm2GsbcFr6RXOucnRmDSDVZjTw7s+udPLds/kD8AdJ15jZc/3NWFID0Xhq9eE4N5vZ+ZJagJuAKUTTFXzQzNrCPucB84Es8DkzuyOkzwCuAYYAtwHnmJlJqgeuA2YAW4APmdnqsM884GuhOBebmY9+4MpWrmZTk8IIAgCTWvzBTldaff4btT+BJtgDvM3MjiYaJXqOpNlEs34uNrPpwOKwjKQjgLnAkcAc4CeScs11lwNnAtPDa05Inw+0mdk04DLg0pBXC3A+cDwwCzhf0sj9/B7Opa4jd88mpZrN8CG1NNXXeLBxJZPOlQ1YZHtYrA0vIxrYM1fLuBY4NXw+BbjRzPaY2SpgBTBL0nig2cyWmJkR1WTi++Tyuhk4UZKAk4FFZtYaak2L2BugnCs7mdCMltY9G0lMbBnKmjZ/sNOVRmrBBkBStaRHgE1Ef/zvB8aZ2XqA8D42bD4BWBPbfW1ImxA+d0/fZx8z6ySaxnpUnry6l+9MSUslLd28efNAvqpzA5JJuWYDMLllCM97zcaVSJ9jo0kaA3ya6B7LS9ub2Sf72tfMssAxkkYAv5F0VL5D9ZRFnvT93SdeviuAKwBmzpz5svXOFUtu+P80Rn3OmTRyKH94ZjNmRtQA4FzxJBmI8xbgj8D/Et247zcze1HSPURNWRsljTez9aGJbFPYbC0wKbbbRGBdSJ/YQ3p8n7WSaoDhQGtIf0u3fe7Zn7I7Vwx7n7NJr2YzqWUouzNdbN6+h7FNDakdx7meJLmyh5rZl81sgZn9KvfqaydJY0KNBklDgLcDTwELgXlhs3lEwYyQPldSvaSpRB0BHghNbe2SZof7MWd02yeX12nAXeG+zh3ASZJGho4BJ4U058pS2vdsINb92QfkdCWQpGZzq6R3mdlt/cx7PHBt6FFWBSwws1slLQEWSJoPPA+cDmBmT0haADxJNO302aEZDuAz7O36fHt4AVwFXC9pBVGNZm7Iq1XSRcCDYbsLzay1n+V3rmiKcc8m3v15xiHeOdMVV74RBNrZe//jK5L2AJmwbGbWnC9jM3sUeG0P6VuAE3vZ5xLgkh7SlwIvu99jZrsJwaqHdVcDV+cro3PlYm+wSe9eysSXRhHwTgKu+PI91NlUzII4V8kyKY8gANBQW82YpnofRcCVRJJpoRcnSXPO7b9iNKNBdN/Guz+7UsjXjNYANAKjw032XP2+GTi4CGVzrmIUoxkNYNLIITy4ui3VYzjXk3wdBP4B+DxRYHk4lr4N+HGahXKu0ux9zibdms2klqEs/Os6MilNQe1cb/Lds/k+8H1JnzUzn7/GuRS91PU5xedsIAo2XQbrXtzFIaMaUz2Wc3H5mtHeZmZ3AX+T9P7u683Mp4V2rkCKdc8mPq+NBxtXTPma0d4M3AW8t4d1Bniwca5AinXP5pBRUbBZvWUHb5g+OtVjOReXrxnt/PD+ieIVx7nKlPYUAznjhzfQWFfNik3b+97YuQJKMhDns8B9ROOj3WtmT6ZeKucqTKYz/edsIJpqYNrYYR5sXNElubKPAH5KNHT/f0haKek36RbLucrS2dVFdZWorkp/NOZpY5tYvqk99eM4F5ck2GSJhqnJAl3ARvaO1OycK4CObBc1RQg0ANPHDWPjtj1s3ZUpyvGcg2QDcW4DHgO+C/wsjG3mnCugTKelOuJz3PSxwwBYsamdGYe0FOWYziW5uj8M3Av8I3CjpG9I6nEgTefc/slku1KdyyZu+tho2MPlG/2+jSuePms2ZnYLcIukVwHvJBpV4F+Ihvt3zhVA9ER/cZrRJo4cQkNtFcu9k4AroiQDcf4q9Ej7PtFYaWcAPhmGcwXUUcThY6qqoh5pHmxcMSW5Z/Mt4OHYRGbOuQLLZIt3zwaiprT7V/rtV1c8fV7dZvagBxrn0pXpLO7AmNPGDmN+5+zzAAAWCUlEQVTd1t207/Yeaa44fNhX58pA1EGgOPdsYG+PtGc37yjaMV1l82DjXBko5j0bgOnjcj3S/OFOVxxJOgicIKkxfP6opO9KOiT9ojlXOTqzRm1V8YLNpJFDqKup8mFrXNEkubovB3ZKOpqoy/NzwHWplsq5ClPsZrSa6ioOHd3oPdJc0SQJNp1mZsApwPfDpGpNfe0kaZKkuyUtk/SEpHNCeoukRZKWh/eRsX3Ok7RC0tOSTo6lz5D0WFj3A0kK6fWSbgrp90uaEttnXjjGcknzkp4Q50qhFDNnTh/nY6S54klydbdLOg/4KPA7SdVAbYL9OoEvmtnhwGzgbElHAOcCi81sOrA4LBPWzQWOBOYAPwnHgqh2dSYwPbzmhPT5QJuZTQMuAy4NebUA5wPHA7OA8+NBzbly05G14gebscNY27aLnR2dRT2uq0xJru4PAXuA+Wa2AZgAfLuvncxsvZk9HD63A8vCvqcA14bNrgVODZ9PAW40sz1mtgpYAcySNB5oNrMloYZ1Xbd9cnndDJwYaj0nA4vMrNXM2oBF7A1QzpWdTLarqM/ZALxy3DDM4NlN3iPNpS/v1R1qFr8ws++a2R8BzOx5M+vXPZvQvPVa4H5gnJmtD3mtB8aGzSYAa2K7rQ1pE8Ln7un77GNmncBWoqkQesure7nOlLRU0tLNmzf35ys5V1DFHK4mZ1pujDRvSnNFkDfYhIc5d0oavr8HkDQM+BXweTPblm/TnoqQJ31/99mbYHaFmc00s5ljxozJUzTn0lXshzohmiK6tlreScAVRZLhanYDj0laBLxU3zazz/W1o6RaokBzg5n9OiRvlDTezNaHJrLc3DhrgUmx3ScC60L6xB7S4/uslVQDDAdaQ/pbuu1zT5/f1LkS6cha0UZ9zqmtrmLq6EYf/dkVRZKr+3fAvxJNM/BQ7JVXuHdyFbDMzL4bW7UQyPUOmwfcEkufG3qYTSXqCPBAaGprlzQ75HlGt31yeZ0G3BXu69wBnCRpZOgYcFJIc64sleKeDURjpHkzmiuGJFMMXCtpCDDZzJ7uR94nAB8jqhU9EtK+QjSw5wJJ84HngdPDcZ6QtAB4kqgn29mxMdk+A1xDNK3B7eEFUTC7XtIKohrN3JBXq6SLgAfDdheaWWs/yu5cUWWKOFNn3GEHNXHb4+vZsaeTxvokDR3O7Z8+ry5J7wX+A6gDpko6huiP9/vy7Wdm/0fP904Aepx8zcwuAS7pIX0pcFQP6bsJwaqHdVcDV+cro3PlorMEzWgAh49vxgye2tDOjEP86QCXniRX9wVEz6q8CGBmjwBTUyyTcxXFzIo+NlrO4eOjHmnL1ufru+PcwCUdQWBrt7SX9exyzu2fzq7o16muyF2fASaMGEJTQ40HG5e6JI20j0v6e6Ba0nTgc8Cf0y2Wc5Ujk+0CKEnNRhKHH9TMUxu8k4BLV5Kr+7NEQ8jsAf4L2AZ8Ps1COVdJMp1RzaYUwQaiprSn1m+jq8sbLFx6kvRG2wl8NbyccwXWkavZlKCDAESdBHZ0ZFnTtpNDRjWWpAzuwJekN9pMoi7LU+Lbm9lr0iuWc5Uj14xWins2EAUbiDoJeLBxaUlyz+YG4J+Bx4CudIvjXOUp5T0biJ61qRI8ub6dOUeNL0kZ3IEvSbDZbGYLUy+JcxWq1MGmobaaqaMbvUeaS1WSYHO+pCuJ5p7Zk0uMjXXmnBuATDbXQaA0zWgQNaU9subFkh3fHfiSBJtPAK8imjAt14xmgAcb5wqg1DUbiILNrY+uZ9vuDM0NSeZGdK5/kgSbo83s1amXxLkKVQ7B5ojQSeCp9e3MmtpSsnK4A1eSq/u+MGWzcy4FHSV+zgb27ZHmXBqS1GzeAMyTtIrono0A867PzhXGS12fa0p3z2Zccz0jh9Z6sHGpSRJs5qReCucqWDk0o0ni8PHNHmxcavq8us3sudwLODn22TlXAOUQbCBqSnt6YztZH7bGpaC/V/dZqZTCuQrWkS39PRuIgs3uTBert+zoe2Pn+qm/V3fpGpWdO0BlOnPD1ZQ22LzqoGhumyfXeVOaK7w+r25J9bHF94Y07xvpXIHkmtFqSvhQJ0TD1gyprWbpap9B3RVekn+lfi2pFsDM1koaDyxKt1jOVY5MV3k0o9VWV3Hc1BaWrNxS0nK4A1OSq/u3wH9LqpY0BbgDOC/NQjlXScqlGQ3gdYeO4pmN29ncvqfvjZ3rhyS90X5GVJP5LfA/wFlmdmfaBXOuUrzUG62Ez9nkvO4VowC4z2s3rsB6DTaSvpB7AQ3AJOARYHZIy0vS1ZI2SXo8ltYiaZGk5eF9ZGzdeZJWSHpa0smx9BmSHgvrfiBJIb1e0k0h/f5Q68rtMy8cY7mkef07Jc4VV7l0fQY46uBmhtXXeFOaK7h8V3dT7DUM+A2wIpbWl2t4+QOh5wKLzWw60SjS5wKE4XDmEk0/PQf4iaTqsM/lwJnA9PDK5TkfaDOzacBlwKUhrxbgfOB4YBbRqNUvBTXnyk2u63NNVelrNjXVVRw/tYUlz3qwcYXV6wgCZvaNgWRsZvfGaxvBKcBbwudrgXuAL4f0G81sD7BK0gpglqTVQLOZLQGQdB1wKnB72OeCkNfNwI9CredkYJGZtYZ9FhEFqP8ayPdxLi2ZbBd11VWESnvJve4Vo1j81CY2bN3NQcMbSl0cd4Aodr19nJmtBwjvY0P6BGBNbLu1IW1C+Nw9fZ99zKwT2AqMypOXc2Up09lV0rlsupt9aHTfZsnKF0pcEncgKX0jcaSn3zTLk76/++x7UOlMSUslLd28eXOigjpXaJlsF7U15fKrGE03MHxIrTeluYIq9hW+MTynQ3jfFNLXEnVAyJkIrAvpE3tI32cfSTXAcKA1T14vY2ZXmNlMM5s5ZsyYAXwt5/ZfR9aoqSqfYFNVJWYf2sKfPdi4Aur1no2kr+fZz8zsov043kJgHvCt8H5LLP2Xkr4LHEzUEeABM8tKapc0G7gfOAP4Ybe8lgCnAXeZmUm6A/hmrFPASfhzQa6MRfdsyqcZDaLnbe54YiNrWncyqWVoqYvjDgD5phjoaTS+ocCniO6N5A02kv6LqDPAaElriXqIfQtYIGk+8DxwOoCZPSFpAfAk0AmcbWbZkNVniHq2DSHqGHB7SL8KuD50Jmgl6s2GmbVKugh4MGx3Ya6zgHPlqLPMmtEAXveK0QAsWbnFg40riHy90b6T+yypCTgH+CRwI/Cd3vaL7f/hXlad2Mv2lwCX9JC+FDiqh/TdhGDVw7qrgav7KqNz5SCTtbJ4xibuleOGMaqxjiXPbuGDMyf1vYNzfch7hYeHMC8GHiUKTMea2ZfNbFO+/ZxzyXVku8ou2Ehi9itG8cflL7z00KlzA5FvBIFvEzVFtQOvNrMLzKytaCVzrkKU4z0bgPe/dgIvbN/DbY+tL3VR3AEg379TXyS6Wf81YJ2kbeHVLsknvHCuQDJlWLMBeOthY3nFmEauuHclZj57pxuYXq9wM6sysyFm1mRmzbFXk5k1F7OQzh3IMp3ld88Goi7Qn37joTyxbps/c+MGrPyucOcqTEcZ9kbLOfW1Exg9rI6f/XFlqYviBrnyvMKdqyDles8GoKG2mnmvm8LdT2/mmY3tpS6OG8Q82DhXYplsV1mNINDdR2cfQkNtFVd67cYNQPle4c5ViEzWyrYZDWBkYx0fnDmJ3/5lHZu27S51cdwgVb5XuHMVIuqNVp7NaDnz3zCVzq4ufnz3ilIXxQ1SHmycK7HcfDbl7JBRjXx09iFcf99zLFvvTz64/ivvK9y5ClCOw9X05AvveCXDh9Ry/sIn/Lkb12/lf4U7d4CLJk8r/1/FEUPr+OeTX8UDq1pZ+NceZ+1wrlflf4U7d4CLnrMp73s2OR86bhKvnjCcb962jB17OktdHDeIeLBxrsQGwz2bnOoq8Y1TjmTjtj388C7vLOCSGxxXuHMHqGyX0WUMima0nGMnj+T0GRP52R9XcvdTPgC8S2bwXOHOHYByw/fXlHnX5+4ueN+RHDG+mbN/+TCPrd1a6uK4QcCDjXMl1BGCzWBpRstprK/hqo/PZOTQOj5xzYOsad1Z6iK5Mje4rnDnDjCd2agL8WBqRssZ29TAtZ88jo7OLB//+QO07ugodZFcGRt8V7hzB5BcM9pgDDYA08Y28bMzZrKmdRdzvncvi5dtLHWRXJkanFe4cweIjs5csBlc92zijj90FL/+x9fT0ljH/GuX8oWbHmHrzkypi+XKjAcb50ooV7OpK+OBOJM4asJwFv7TG/js26Zxy1/X8bbv3MPP7l3Jro5sqYvmysTgvsKdG+Qyg/ieTXd1NVV88aTDuOXsEzh8fDOX3LaMN/773Vz5x5W8sH1PqYvnSqym1AVIk6Q5wPeBauBKM/tWiYvk3D4G+z2bnhw1YTi/+NTxPLCqle8vfoaLf7eMi3+3jMktQzl28giOnjSCw8Y1MX1cE6OH1SEN3iZEl9wBG2wkVQM/Bt4BrAUelLTQzJ4sbcmc26sjO/jv2fRm1tQWbvjUbB5d+yJLnt3Cw8+38adnt/DbR/aOq9bSWMf44Q2MHlbP6GH1jBxaS0NtNQ21VTTUVlNfW01DTfgc3uPrG2qqqa+tor6miqoqUSVRJaiSUHjPpXlQK60DNtgAs4AVZrYSQNKNwCmABxtXNjKdB17NprvXTBzBayaOAMDM2NS+h2c2trN843aWb9rOxm27eWH7HlZs2k7bzg52Z7J0pTSodFUsAOWCUXXV3s9J4lFfmyQJan3n0Xc5BpsDOdhMANbEltcCx8c3kHQmcCbA5MmTi1cy54IRQ+t496vHM6apvtRFKQpJjGtuYFxzA2+cPqbHbcyMTNbY3ZlldybLnkwXezqz7M50sTsTe4+ldXR20WWGGXRZNARQtLz3c5cRlmNpXft+7ktfWySZecH6yCVZHn0HrGJ5KOF2B3Kw6elnsc+P0cyuAK4AmDlzpk/Q4YrusIOa+PFHji11McqKJOpqRF1NFc0NtaUujuvDNxNud+DW3aOazKTY8kTAJ+FwzrkSOJCDzYPAdElTJdUBc4GFJS6Tc85VpAO2Gc3MOiX9E3AHUdfnq83siRIXyznnKtIBG2wAzOw24LZSl8M55yrdgdyM5pxzrkx4sHHOOZc6DzbOOedS58HGOedc6mRJHletAJI2A88VKLvRwAsFyutA4ueld35ueubnpWfldF4OMbOeh4OI8WCTAklLzWxmqctRbvy89M7PTc/8vPRsMJ4Xb0ZzzjmXOg82zjnnUufBJh1XlLoAZcrPS+/83PTMz0vPBt158Xs2zjnnUuc1G+ecc6nzYOOccy51HmwKTNKXJJmk0bG08yStkPS0pJNLWb5SkPRtSU9JelTSbySNiK2r9HMzJ3z3FZLOLXV5SkXSJEl3S1om6QlJ54T0FkmLJC0P7yNLXdZSkFQt6S+Sbg3Lg+68eLApIEmTgHcAz8fSjiCaS+dIYA7wE0nVpSlhySwCjjKz1wDPAOeBn5vwXX8MvBM4AvhwOCeVqBP4opkdDswGzg7n4lxgsZlNBxaH5Up0DrAstjzozosHm8K6DPgX9p1++hTgRjPbY2argBXArFIUrlTM7E4z6wyL9xHNmgp+bmYBK8xspZl1ADcSnZOKY2brzezh8Lmd6A/rBKLzcW3Y7Frg1NKUsHQkTQTeDVwZSx5058WDTYFIeh/wNzP7a7dVE4A1seW1Ia1SfRK4PXyu9HNT6d+/R5KmAK8F7gfGmdl6iAISMLZ0JSuZ7xH9E9sVSxt05+WAnjyt0CT9L3BQD6u+CnwFOKmn3XpIO+D6m+c7N2Z2S9jmq0TNJTfkduth+wPu3ORR6d//ZSQNA34FfN7Mtkk9naLKIek9wCYze0jSW0pdnoHwYNMPZvb2ntIlvRqYCvw1/HJMBB6WNIvov9VJsc0nAutSLmrR9XZuciTNA94DnGh7H+6qiHOTR6V//31IqiUKNDeY2a9D8kZJ481svaTxwKbSlbAkTgDeJ+ldQAPQLOkXDMLz4s1oBWBmj5nZWDObYmZTiP6IHGtmG4CFwFxJ9ZKmAtOBB0pY3KKTNAf4MvA+M9sZW1Xp5+ZBYLqkqZLqiDpLLCxxmUpC0X9pVwHLzOy7sVULgXnh8zzglmKXrZTM7Dwzmxj+rswF7jKzjzIIz4vXbFJmZk9IWgA8SdSEdLaZZUtcrGL7EVAPLAo1v/vM7KxKPzdm1inpn4A7gGrgajN7osTFKpUTgI8Bj0l6JKR9BfgWsEDSfKJenqeXqHzlZtCdFx+uxjnnXOq8Gc0551zqPNg455xLnQcb55xzqfNg45xzLnUebJxzzqXOg42rSJKmSHq81OXoiaRrJJ2WUt7b08jXub54sHGuSCppRGvnuvNg4yqepEPDXCHH9bDuHkmXSbo3zLVynKRfh3lELo5t91FJD0h6RNJPc4FF0nZJF0q6H3hdt7w/LelBSX+V9CtJQ2Or3y7pj5KeCeNjIenjkn4U2//W3HhZ4TiXhLzukzQupE+VtCQc56LYvsMkLZb0sKTHJJ0S0qeE7/mzMK/MnZKGhHXTJP1vOMbDkl4R0v855P+opG8M8MfhDlAebFxFk3QY0XhcnzCzB3vZrMPM3gT8J9GwIGcDRwEflzRK0uHAh4ATzOwYIAt8JOzbCDxuZseb2f91y/fXZnacmR1NNKT+/Ni6KcCbiYaW/09JDX18lUaikRmOBu4FPh3Svw9cbmbHARti2+8G/s7MjgXeCnxHe0e9nA782MyOBF4EPhDSbwjpRwOvB9ZLOilsPws4Bpgh6U19lNVVIB+uxlWyMUTB4wN9DBOTG6/sMeCJ3NDuklYSDaT5BmAG8GD4ez2EvQMjZomCWU+OCrWjEcAwomFrchaYWRewPBznVX18lw7g1vD5IaJJ/CAaBiYXLK4HLg2fBXwzBIYuoqkNxoV1q8zskVheUyQ1ARPM7DcAZrY7nIOTiEY7/0vYfhhR8Lm3j/K6CuPBxlWyrUTzyZwAPAEg6edEc6msM7N3he32hPeu2Ofccg3RH+5rzey8Ho6xO894b9cAp5rZXyV9HHhLbF33caSMaPy4eGtEvLaTiY2mnWXf3+2exqT6CFGwnWFmGUmrY/nFv2OWKHj2Nta/gH8zs5/2st45wJvRXGXrIJrh8AxJfw9gZp8ws2NigSaJxcBpksbCS/PDH5Jgvyaipqha9ja75ZwuqSrcFzkUeBpYDRwT0ieRbFbTPxGNFky3YwwnmiclI+mtQN7ymtk2YK2kUwHCSN1DiWpjn1Q0Dw2SJuTOg3NxXrNxFc3MdoQb8Isk7chN9NbPPJ6U9DXgTklVQIbovs5zfez6r0SzUT5H1ETXFFv3NPAHoqats8xst6Q/AavCto8DDyco3jnALyWdw77NeTcA/yNpKfAI8FSCvD4G/FTShUTf8XQzuzPcs1oSmhC3Ax9lEMyv4orLR312zjmXOm9Gc845lzoPNs4551LnwcY551zqPNg455xLnQcb55xzqfNg45xzLnUebJxzzqXu/wO2dsjkd2fsfwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "!abundance-dist-single.py -M 1e9 -k 21 -s ERX168346.corrected.trimmed.fq ERX168346.dist\n",
    "\n",
    "dist1 = numpy.loadtxt('ERX168346.dist', skiprows=1, delimiter=',')\n",
    "plot(dist1[:,0], dist1[:,1])\n",
    "axis(xmax=50)\n",
    "title('k-mer abundance histogram')\n",
    "xlabel('k-mer abundance')\n",
    "ylabel('N k-mers with that abundance')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So we've managed to reduce the number of low abundance k-mers in our sample using sketching.\n",
    "\n",
    "* just to finish up, let's see if our QC has improved our sketch similarity between the sample and the reference genome:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[K== This is sourmash version 2.0.0. ==\n",
      "\u001b[K== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==\n",
      "\n",
      "\u001b[Ksetting num_hashes to 0 because --scaled is set\n",
      "\u001b[Kcomputing signatures for files: ERX168346.corrected.trimmed.fq\n",
      "\u001b[KComputing signature for ksizes: [31]\n",
      "\u001b[KComputing only nucleotide (and not protein) signatures.\n",
      "\u001b[KComputing a total of 1 signature(s).\n",
      "\u001b[K... reading sequences from ERX168346.corrected.trimmed.fq\n",
      "\u001b[Kcalculated 1 signatures for 250557 sequences in ERX168346.corrected.trimmed.fq\n",
      "\u001b[Ksaved 1 signature(s). Note: signature license is CC0.\n",
      "\u001b[K== This is sourmash version 2.0.0. ==\n",
      "\u001b[K== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==\n",
      "\n",
      "\u001b[Kloaded query: ERX168346.corrected.trimmed.fq... (k=31, DNA)\n",
      "\u001b[Kloaded 1 databases.                                                            \n",
      "\n",
      "\n",
      "overlap     p_query p_match\n",
      "---------   ------- -------\n",
      "4.4 Mbp       97.8%   91.7%    FKBP01000001.1 Enterobacter cloacae s...\n",
      "\u001b[Kfound less than 10.0 kbp in common. => exiting\n",
      "\n",
      "found 1 matches total;\n",
      "the recovered matches hit 97.8% of the query\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# recompute the sketch for our sample, this time using the QC'd reads\n",
    "!sourmash compute --scaled 1000 -k 31 ERX168346.corrected.trimmed.fq --o ERX168346.sketch\n",
    "\n",
    "# compare the sketch against the reference genomes\n",
    "!sourmash gather -k 31 ERX168346.sketch genbank-k31.lca.json.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Great - we now have increased the amount of query matching the reference. In other words, the similarity between our sample and the reference genome has increased!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "\n",
    "Let's move on to the next stage of the workflow: [resistome profiling](r4.3.Resistome-profiling.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:genome-sketching] *",
   "language": "python",
   "name": "conda-env-genome-sketching-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
